gateway/mw_graphql_complexity_test.go

```diff
@@ -48,67 +49,215 @@ func TestGraphQLComplexityMiddleware_DepthLimitExceeded(t *testing.T) {
-		failReason := m.DepthLimitExceeded(req, accessDef, countriesSchema)
-		assert.Equal(t, ComplexityFailReasonDepthLimitExceeded, failReason)
-	})
+	session := user.NewSessionState()
+	session.MaxQueryDepth = 3
+	ctxSetSession(httpReq, session, "", false)
 
-	t.Run("should respect unlimited specific field depth limit and not exceed", func(t *testing.T) {
-		accessDef.FieldAccessRights = []user.FieldAccessDefinition{
-			{
-				TypeName:  "Query",
-				FieldName: "countries",
-				Limits: user.FieldLimits{
-					MaxQueryDepth: -1,
+	cases := []struct {
+		name   string
+		rights map[string]user.AccessDefinition
+		result int
+	}{
+		{
+			name:   "should fail when no access rights for the given spec",
+			rights: map[string]user.AccessDefinition{"any": {}},
+			result: http.StatusInternalServerError,
+		},
+		{
+			name: "should take limit from access rights and fail",
+			rights: map[string]user.AccessDefinition{
+				apiSpec.APIID: {
+					Limit: &user.APILimit{MaxQueryDepth: 1},
 				},
 			},
-		}
-
-		failReason := m.DepthLimitExceeded(req, accessDef, countriesSchema)
-		assert.Equal(t, ComplexityFailReasonNone, failReason)
-	})
-
-	t.Run("should respect higher specific field depth limit and not exceed", func(t *testing.T) {
-		accessDef.FieldAccessRights = []user.FieldAccessDefinition{
-			{
-				TypeName:  "Query",
-				FieldName: "countries",
-				Limits: user.FieldLimits{
-					MaxQueryDepth: 10,
+			result: http.StatusForbidden,
+		},
+		{
+			name: "should take field limits from access rights and proceed",
+			rights: map[string]user.AccessDefinition{
+				apiSpec.APIID: {
+					FieldAccessRights: []user.FieldAccessDefinition{
+						{
+							TypeName: "Query", FieldName: "countries",
+							Limits: user.FieldLimits{MaxQueryDepth: 3},
+						},
+					},
 				},
 			},
-		}
-
-		failReason := m.DepthLimitExceeded(req, accessDef, countriesSchema)
-		assert.Equal(t, ComplexityFailReasonNone, failReason)
-	})
-
-	t.Run("should respect lower specific field depth limit and exceed", func(t *testing.T) {
-		accessDef.Limit.MaxQueryDepth = 100
-		accessDef.FieldAccessRights = []user.FieldAccessDefinition{
-			{
-				TypeName:  "Query",
-				FieldName: "countries",
-				Limits: user.FieldLimits{
-					MaxQueryDepth: 1,
+			result: http.StatusOK,
+		},
+		{
+			name: "should take limit from session and fail when access rights with not matching field and empty limit",
+			rights: map[string]user.AccessDefinition{
+				apiSpec.APIID: {
+					FieldAccessRights: []user.FieldAccessDefinition{
+						{
+							TypeName: "Query", FieldName: "continents",
+							Limits: user.FieldLimits{MaxQueryDepth: 4},
+						},
+					},
 				},
 			},
-		}
+			result: http.StatusForbidden,
+		},
+	}
 
-		failReason := m.DepthLimitExceeded(req, accessDef, countriesSchema)
-		assert.Equal(t, ComplexityFailReasonDepthLimitExceeded, failReason)
-	})
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			session.SetAccessRights(tc.rights)
```

Comment:
@spetrunin session changed. I think it should be `session.AccessRights=tc.rights`